exchange/bitswap/bitswap.go

```diff
@@ -320,43 +321,27 @@ func (bs *Bitswap) sendWantlistToProviders(ctx context.Context, entries []wantli
 }
 
 // TODO(brian): handle errors
-func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg.BitSwapMessage) (
-	peer.ID, bsmsg.BitSwapMessage) {
+func (bs *Bitswap) ReceiveMessage(ctx context.Context, p peer.ID, incoming bsmsg.BitSwapMessage) error {
 	defer log.EventBegin(ctx, "receiveMessage", p, incoming).Done()
 
-	if p == "" {
-		log.Debug("Received message from nil peer!")
-		// TODO propagate the error upward
-		return "", nil
-	}
-	if incoming == nil {
-		log.Debug("Got nil bitswap message!")
-		// TODO propagate the error upward
-		return "", nil
-	}
-
 	// This call records changes to wantlists, blocks received,
 	// and number of bytes transfered.
 	bs.engine.MessageReceived(p, incoming)
 	// TODO: this is bad, and could be easily abused.
 	// Should only track *useful* messages in ledger
 
+	var keys []u.Key
 	for _, block := range incoming.Blocks() {
+		log.Debugf("got block %s from %s", block, p)
 		hasBlockCtx, cancel := context.WithTimeout(ctx, hasBlockTimeout)
 		if err := bs.HasBlock(hasBlockCtx, block); err != nil {
-			log.Debug(err)
+			return fmt.Errorf("ReceiveMessage HasBlock error: %s", err)
 		}
 		cancel()
-	}
-
-	var keys []u.Key
-	for _, block := range incoming.Blocks() {
 		keys = append(keys, block.Key())
 	}
-	bs.cancelBlocks(ctx, keys)
 
-	// TODO: consider changing this function to not return anything
-	return "", nil
+	return bs.cancelBlocks(ctx, keys)
```

Comment:
this error probably should not be returned by ReceiveMessage, but logged or otherwise handled. AFAIK, if this fails, it does not mean a ReceiveMessage failure.